src/back/space/src/service/service.cpp

```diff
@@ -0,0 +1,397 @@
+	else
+		throw errors::NotFound{};
+	return {};
+}
+
+model::Space Service::GetByKey(const std::string& key, const std::string& userId) {
+	const auto space = _repo.Space().SelectByKey(key);
+	const auto isUserInside = _repo.SpaceUser().IsUserInside(space.id, userId);
+	if (isUserInside || space.requestsAllowed)
+		return space;
+	else
+		throw errors::NotFound{};
+	return {};
+}
+
+model::Space Service::GetByLink(const boost::uuids::uuid& link) {
+	boost::uuids::uuid spaceUuid = _repo.SpaceLink().GetSpaceId(link);
+	if (!spaceUuid.is_nil())
+		return _repo.Space().SelectById(spaceUuid);
+	return {};
+}
+
+bool Service::IsLinkExpired(const boost::uuids::uuid& link) {
+	model::SpaceLink linkEntity = _repo.SpaceLink().SelectById(link);
+	const auto p1 = std::chrono::system_clock::now();
+	const auto now = std::chrono::duration_cast<std::chrono::seconds>(p1.time_since_epoch()).count();
+	if (linkEntity.expiredAt > now)
+		return false;
+	return true;
+}
+
+void Service::InviteByLink(const std::string& creatorId, const boost::uuids::uuid& link) {
+	// todo - is some business logic needed for invitation by link like it was for invitation by login?
+	model::SpaceLink linkEntity = _repo.SpaceLink().SelectById(link);
+	const auto spaceUuid = linkEntity.spaceId;
+	const auto userId = creatorId;
+	const auto role = Role::Type::Unknown;
+	// todo - how to check for duplicates here (invitation that already was inserted)?
+	_repo.SpaceInvitation().Insert(spaceUuid, userId, role, creatorId);
+}
+
+bool Service::CanDeleteUser(const std::string& requestUserId, const boost::uuids::uuid& spaceId, const std::string& userId) {
+	const auto user = _repo.SpaceUser().GetByIds(spaceId, userId);
+
+	if (user.isOwner)
+		return false;
+
+	if (userId == requestUserId)
+		return true;
+	else {
+		const auto isRequestUserAdmin = _repo.SpaceUser().IsAdmin(spaceId, requestUserId);
+		if (isRequestUserAdmin)
+			return true;
+	}
+	return false;
+}
+
+void Service::DeleteUser(const boost::uuids::uuid& spaceId, const std::string& userId) {
+	_repo.SpaceUser().Delete(spaceId, userId);
+}
+
+bool Service::CanUpdateUser(bool isRoleMode, bool isOwner, const boost::uuids::uuid& spaceUuid, const std::string& userId, const std::string& headerUserId) {
+	const auto headerUser = _repo.SpaceUser().GetByIds(spaceUuid, headerUserId);
+
+	const auto user = _repo.SpaceUser().GetByIds(spaceUuid, userId);
+
+	if (isOwner) {
+		if (user.isOwner)
+			return false;
+		if (!headerUser.isOwner)
+			return false;
+		return true;
+	} else {
+		if (isRoleMode)
+			if (headerUser.role == Role::Type::Admin)
+				if (!user.isOwner)
+					return true;
+	}
+	return false;
+}
+
+void Service::UpdateUser(bool isRoleMode, const Role::Type& role, bool isOwner, const boost::uuids::uuid& spaceUuid, const std::string& userId, const std::string& headerUserId) {
+	const auto headerUser = _repo.SpaceUser().GetByIds(spaceUuid, headerUserId);
+
+	const auto user = _repo.SpaceUser().GetByIds(spaceUuid, userId);
```

Comment:
Если тебе нужно менять поля, то значит это не константная переменная. 
```suggestion
	auto user = _repo.SpaceUser().GetByIds(spaceUuid, userId);

	// У владельца ничего менять нельзя
	if (user.isOwner)
		return false;
```